### PR TITLE
Remove TempDir references and os path manipulation in tests

### PR DESCRIPTION
**Pull request by BenWilson2**
Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

## What changes are proposed in this pull request?

Remove os.path refs in Diviner, Pmdarima, and Prophet test suites. 
Cleanup TempDir references.

## How is this patch tested?

Unit tests

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [x] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/diviner/test_diviner_model_export.py [hunk 1]
```diff
@@ -304,13 +303,13 @@ def test_diviner_model_save_persists_specified_conda_env_in_mlflow_model_directo
...
-    assert os.path.exists(saved_conda_env_path)
-    assert saved_conda_env_path != diviner_custom_env
+    saved_conda_env_path = pathlib.Path(model_path).joinpath(pyfunc_conf[pyfunc.ENV])
+    assert saved_conda_env_path.exists()
+    assert str(saved_conda_env_path) != diviner_custom_env
 
     with open(diviner_custom_env, "r") as f:
         diviner_custom_env_parsed = yaml.safe_load(f)
-    with open(saved_conda_env_path, "r") as f:
+    with open(str(saved_conda_env_path), "r") as f:
```
Comment (harupy) [inline tests/diviner/test_diviner_model_export.py:312]:
```suggestion
    with open(saved_conda_env_path, "r") as f:
```

`open` should work without `str`.

File: tests/diviner/test_diviner_model_export.py [hunk 2]
```diff
@@ -74,9 +72,9 @@ def grouped_pmdarima(diviner_data):
 
 @pytest.fixture
 def diviner_custom_env(tmpdir):
-    conda_env = os.path.join(str(tmpdir), "conda_env.yml")
-    _mlflow_conda_env(conda_env, additional_pip_deps=["diviner"])
-    return conda_env
+    conda_env = pathlib.Path(tmpdir).joinpath("conda_env.yml")
+    _mlflow_conda_env(str(conda_env), additional_pip_deps=["diviner"])
+    return str(conda_env)
```
Comment (harupy) [inline tests/diviner/test_diviner_model_export.py:77]:
```suggestion
    conda_env = tmp_path.joinpath("conda_env.yml")
    _mlflow_conda_env(conda_env, additional_pip_deps=["diviner"])
    return conda_env
```

Can you try this?

File: tests/diviner/test_diviner_model_export.py [hunk 3]
```diff
@@ -198,27 +196,28 @@ def test_diviner_pyfunc_group_predict_pmdarima(grouped_pmdarima, model_path, div
...
-                mlflow_model = Model.load(path)
-                assert signature == mlflow_model.signature
-                if example is None:
-                    assert mlflow_model.saved_input_example_info is None
-                else:
-                    r_example = _read_example(mlflow_model, path).copy(deep=False)
-                    # NB: datetime values are implicitly cast, so this needs to be reverted.
-                    r_example["ds"] = pd.to_datetime(r_example["ds"], format=DS_FORMAT)
-                    np.testing.assert_array_equal(r_example, example)
+            idx += 1
```
Comment (harupy) [inline tests/diviner/test_diviner_model_export.py:207]: Can we parametrize this test for `signature` and `example` and remove `idx`?

File: tests/diviner/test_diviner_model_export.py [hunk 4]
```diff
@@ -321,9 +320,8 @@ def test_diviner_model_save_persists_requirements_in_mlflow_model_directory(
     mlflow.diviner.save_model(
         diviner_model=grouped_pmdarima, path=model_path, conda_env=diviner_custom_env
     )
-
-    saved_pip_req_path = os.path.join(model_path, "requirements.txt")
-    _compare_conda_env_requirements(diviner_custom_env, saved_pip_req_path)
+    saved_pip_req_path = pathlib.Path(model_path).joinpath("requirements.txt")
```
Comment (harupy) [inline tests/diviner/test_diviner_model_export.py:323]:
```suggestion
    saved_pip_req_path = model_path.joinpath("requirements.txt")
```

`model_path` is already a `Path` object, right?

File: tests/prophet/test_prophet_model_export.py [hunk 5]
```diff
@@ -202,49 +202,48 @@ def test_model_load_from_remote_uri_succeeds(prophet_model, model_path, mock_s3_
...
-
-
-def test_log_model_calls_register_model(prophet_model):
+def test_prophet_log_model(prophet_model, tmpdir):
+    for should_start_run in [False, True]:
+        try:
+            if should_start_run:
+                mlflow.start_run()
+            artifact_path = "prophet"
+            conda_env = pathlib.Path(tmpdir).joinpath("conda_env.yaml")
```
Comment (harupy) [inline tests/prophet/test_prophet_model_export.py:211]:
```suggestion
            conda_env = tmp_path.joinpath("conda_env.yaml")
```

Can we replace `tmpdir` with `tmp_path`?

**Comment by harupy** [comment]
@BenWilson2 Left a few initial comments!

File: tests/prophet/test_prophet_model_export.py [hunk 6]
```diff
@@ -348,7 +343,7 @@ def test_log_model_with_extra_pip_requirements(prophet_model, tmpdir):
     # Constraints file
     with mlflow.start_run():
         mlflow.prophet.log_model(
-            prophet_model.model, "model", extra_pip_requirements=[f"-c {req_file.strpath}", "b"]
+            prophet_model.model, "model", extra_pip_requirements=[f"-c {str(req_file)}", "b"]
```
Comment (harupy) [inline tests/prophet/test_prophet_model_export.py:346]:
```suggestion
            prophet_model.model, "model", extra_pip_requirements=[f"-c {req_file}", "b"]
```

`str` here can be removed.

File: tests/pmdarima/test_pmdarima_model_export.py [hunk 7]
```diff
@@ -114,54 +114,51 @@ def test_pmdarima_autoarima_pyfunc_save_and_load(auto_arima_model, model_path):
...
+    assert signature == mlflow_model.signature
+    if example is None:
+        assert mlflow_model.saved_input_example_info is None
+    else:
+        r_example = _read_example(mlflow_model, model_path).copy(deep=False)
+        np.testing.assert_array_equal(r_example, example)
 
 
+@pytest.mark.parametrize("_signature", [True, False])
+@pytest.mark.parametrize("_example", [True, False])
```
Comment (harupy) [inline tests/pmdarima/test_pmdarima_model_export.py:140]:
```suggestion
@pytest.mark.parametrize("use_signature", [True, False])
@pytest.mark.parametrize("use_example", [True, False])
```

Can we rename the parameters?

File: tests/prophet/test_prophet_model_export.py (hunk 6 above)
Comment (BenWilson2) [reply tests/prophet/test_prophet_model_export.py:346]: This one still needs a string
```
TypeError: `pip_requirements` must be either a string path to a pip requirements file on the local filesystem or an iterable of pip requirement strings, but got `<class 'pathlib.PosixPath'>`
```

Comment (BenWilson2) [reply tests/prophet/test_prophet_model_export.py:346]: We could fix it by updating https://github.com/mlflow/mlflow/blob/7db74ace1aa7e52e271cd00121d25dbafd14386f/mlflow/utils/environment.py#L172-L214 but maybe we do that in another PR?

File: tests/pmdarima/test_pmdarima_model_export.py (hunk 7 above)
Comment (BenWilson2) [reply tests/pmdarima/test_pmdarima_model_export.py:140]: great idea! Changed!

File: tests/prophet/test_prophet_model_export.py (hunk 6 above)
Comment (harupy) [reply tests/prophet/test_prophet_model_export.py:348]: I think `f"-c {req_file}"` is equivalent to `f"-c {str(req_file)}"` if `req_file` is a pathlib.Path object.

```python
>>> import pathlib
>>> p = pathlib.Path("foo")
>>> f"-c {p}"
'-c foo'
>>> f"-c {str(p)}"
'-c foo'
>>> f"-c {p}" == f"-c {str(p)}"
True
```

Comment (BenWilson2) [reply tests/prophet/test_prophet_model_export.py:348]: Sorry about that @harupy ! I had accidentally changed the one on top (which wasn't embedded in the f-string). Changing those 12 elements across the 3 tests now!

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**